go_covert_lib/controller/processor/symmetricEncryption/symmetricEncryption.go

```diff
@@ -0,0 +1,237 @@
+const padIndex = 7
+
+type SymmetricEncryption struct {
+	algorithm string
+	mode      string
+	key       []byte
+}
+
+func (c *SymmetricEncryption) Process(data []byte) ([]byte, error) {
+	data = Pad(data)
+	algorithm := c.algorithm
+
+	var (
+		blockSize int
+		block     cipher.Block
+		err       error
+	)
+
+	// based on the users choice of symmetric algorithm create a cipher
+	switch algorithm {
+	case "Advanced Encryption Standard (AES)":
+		block, err = aes.NewCipher(c.key)
+		blockSize = 16
+	case "Data Encryption Standard (DES)":
+		block, err = des.NewCipher(c.key)
+		blockSize = 8
+	case "Triple Data Encryption Standard (3DES)":
+		block, err = des.NewTripleDESCipher(c.key)
+		blockSize = 8
+	default:
+		return nil, errors.New("Undefined algorithm selected")
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	// based on the users choice of the mode of operation encrypt in that mode
+	var cipherText []byte
+	selectedMode := c.mode
+	switch selectedMode {
+	case "Galois Counter Mode (GCM)":
+		cipherText = GCMEncrypter(block, data)
+		if cipherText == nil {
+			return nil, errors.New("Unable to encrypt in Galosis 
+									Counter Mode (GCM)")
+		}
+	case "Cipher Block Chaining (CBC)":
+		cipherText = CBCEncrypter(block, data, blockSize)
+	case "Cipher Feedback (CFB)":
+		cipherText = CFBEncrypter(block, data, blockSize)
+	case "Counter (CTR)":
+		cipherText = CTREncrypter(block, data, blockSize)
+	default:
+		return nil, errors.New("Undefined mode selected")
+	}
+
+	return cipherText[:], nil
+}
+
+func GCMEncrypter(block cipher.Block, data []byte) []byte {
+	nonce := make([]byte, 12)
+
+	aesgcm, err := cipher.NewGCM(block)
+	if err != nil {
+		return nil
+	}
+
+	cipherText := aesgcm.Seal(nil, nonce, data, nil)
+
+	return cipherText[:]
+}
+
+func CBCEncrypter(block cipher.Block, data []byte, blockSize int) []byte {
+	cipherText := make([]byte, blockSize+len(data))
+	iv := cipherText[:blockSize]
+
+	mode := cipher.NewCBCEncrypter(block, iv)
+	mode.CryptBlocks(cipherText[blockSize:], data)
+
+	return cipherText[:]
+}
+
+func CFBEncrypter(block cipher.Block, data []byte, blockSize int) []byte {
+	cipherText := make([]byte, blockSize+len(data))
+	iv := cipherText[:blockSize]
+
+	stream := cipher.NewCFBEncrypter(block, iv)
+	stream.XORKeyStream(cipherText[blockSize:], data)
+
+	return cipherText[:]
+}
+
+func CTREncrypter(block cipher.Block, data []byte, blockSize int) []byte {
+	cipherText := make([]byte, blockSize+len(data))
+	iv := cipherText[:blockSize]
+
+	stream := cipher.NewCTR(block, iv)
+	stream.XORKeyStream(cipherText[blockSize:], data)
+
+	return cipherText[:]
+}
+
+func Pad(data []byte) []byte {
+	// padding is done by having 7 zero's at the start followed by the length 
+	// of the original message, the contents of the message, followed by a 
+	// series of zero's to fill the block size requirements of the symmetric  
+	// encryption algorithm 
+	lenOfData := len(data)
+	data = append(data, 0)
+	copy(data[1:], data)
```

Comment:
Again. If you put the length into a byte, then if there is more than 255 bytes this will fail. You will only get the start of the message.

Reply:
Use the "encoding/binary" library to convert a 64 byte int to a byte array. 
len(data) returns a 64 bit int on 64 bit systems and 32 bit int on 32 bit systems. To make it compatible with both we always use an 8 byte length.